iree/tools/benchmark_module_main.cc

```diff
@@ -155,26 +155,33 @@ Status Run(::benchmark::State& state) {
   return OkStatus();
 }
 
-void BM_RunModule(benchmark::State& state) {
+void BM_RunModule(benchmark::State& state, const std::string& function_name) {
   // Delegate to a status-returning function so we can use the status macros.
-  IREE_CHECK_OK(Run(state));
+  IREE_CHECK_OK(RunFunction(state, function_name));
 }
 
-BENCHMARK(BM_RunModule)
-    // By default only the main thread is included in CPU time. Include all the
-    // threads instead.
-    ->MeasureProcessCPUTime()
-    // To make single and multi-threaded benchmarks more comparable, use the
-    // wall time to determine how many iterations to run.
-    // See https://github.com/google/benchmark#cpu-timers,
-    ->UseRealTime()
-    // Report timing in milliseconds, which is the general order of magnitude of
-    // model runs. The benchmark framework will print with precision between 0
-    // and 3 places after the decimal while aiming for three significant digits.
-    // If we end up wanting precision beyond microseconds, we can make this
-    // setting configurable with a custom command line flag.
-    ->Unit(benchmark::kMillisecond);
-
 }  // namespace
 
+extern "C" void RegisterModuleBenchmarks() {
+  auto function_name = absl::GetFlag(FLAGS_entry_function);
+  auto benchmark_name = "BM_" + function_name;
+  benchmark::RegisterBenchmark(benchmark_name.c_str(),
+                               [function_name](benchmark::State& state) {
+                                 BM_RunModule(state, function_name);
+                               })
+      // By default only the main thread is included in CPU time. Include all
+      // the threads instead.
+      ->MeasureProcessCPUTime()
+      // To make single and multi-threaded benchmarks more comparable, use the
+      // wall time to determine how many iterations to run. See
+      // https://github.com/google/benchmark#cpu-timers,
+      ->UseRealTime()
+      // Report timing in milliseconds, which is the general order of magnitude
+      // of model runs. The benchmark framework will print with precision
+      // between 0 and 3 places after the decimal while aiming for three
+      // significant digits. If we end up wanting precision beyond microseconds,
+      // we can make this setting configurable with a custom command line flag.
+      ->Unit(benchmark::kMillisecond);
+}
+
```

Comment:
```suggestion
static void RegisterModuleBenchmarks() {
  auto function_name = absl::GetFlag(FLAGS_entry_function);
  auto benchmark_name = "BM_" + function_name;
  benchmark::RegisterBenchmark(benchmark_name.c_str(),
                               [function_name](benchmark::State& state) {
                                 BM_RunModule(state, function_name);
                               })
      // By default only the main thread is included in CPU time. Include all
      // the threads instead.
      ->MeasureProcessCPUTime()
      // To make single and multi-threaded benchmarks more comparable, use the
      // wall time to determine how many iterations to run. See
      // https://github.com/google/benchmark#cpu-timers,
      ->UseRealTime()
      // Report timing in milliseconds, which is the general order of magnitude
      // of model runs. The benchmark framework will print with precision
      // between 0 and 3 places after the decimal while aiming for three
      // significant digits. If we end up wanting precision beyond microseconds,
      // we can make this setting configurable with a custom command line flag.
      ->Unit(benchmark::kMillisecond);
}

extern "C" int main(int argc, char** argv) {
  ::benchmark::Initialize(&argc, argv);
  iree::InitializeEnvironment(&argc, &argv);
  RegisterModuleBenchmarks();
  ::benchmark::RunSpecifiedBenchmarks();
  return 0;
}
```

Reply:
SGTM! I added an explicit 'main' in 'iree-benchmark-module' also keept the initialization library.